### PR TITLE
[wifi remote]: Add example without `esp_hosted`

### PR DESCRIPTION
**Pull request by david-cermak**
### Task list

- [x] Cleanup git history
- [x] Document and cleanup example README's
- [x] CI for examples
- [x] Bump the version of `esp_wifi_remote`

### CODE REVIEW

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_server.cpp [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+const unsigned char ca_crt[] = "-----BEGIN CERTIFICATE-----\n" CONFIG_ESP_WIFI_REMOTE_EPPP_CLIENT_CA "\n-----END CERTIFICATE-----";
+const unsigned char crt[] = "-----BEGIN CERTIFICATE-----\n" CONFIG_ESP_WIFI_REMOTE_EPPP_SERVER_CRT "\n-----END CERTIFICATE-----";
+const unsigned char key[] = "-----BEGIN RSA PRIVATE KEY-----\n" CONFIG_ESP_WIFI_REMOTE_EPPP_SERVER_KEY "\n-----END RSA PRIVATE KEY-----";
+
+}
+
+using namespace server;
+
+class RpcInstance {
```
Comment (gabsuren) [inline components/esp_wifi_remote/eppp/wifi_remote_rpc_server.cpp:34]: Overall missing @briefs

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+        if (mutex) {
+            vSemaphoreDelete(mutex);
+        }
+        if (events) {
+            vEventGroupDelete(events);
+        }
+    }
+    SemaphoreHandle_t mutex{nullptr};
+    EventGroupHandle_t events{nullptr};
+    const int request = 1;
```
Comment (gabsuren) [inline components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp:69]: shouldn't be better private?

File: components/esp_wifi_remote/wifi_apps/roaming_app/src/Kconfig.roaming
Comment (gabsuren) [inline components/esp_wifi_remote/wifi_apps/roaming_app/src/Kconfig.roaming:1]: Empty?

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    uint32_t size;
+} __attribute((__packed__));
+
+template<typename T>
+struct RpcData {
+    RpcHeader head;
+    T value_{};
+    explicit RpcData(api_id id) : head{id, sizeof(T)} {}
+
+    uint8_t *value()
```
Comment (euripedesrocha) [inline components/esp_wifi_remote/eppp/wifi_remote_rpc_impl.hpp:44]: This could be non-member `operator*`

Comment (euripedesrocha) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_impl.hpp:44]: Or hidden friend

**Review by euripedesrocha — COMMENTED** [review]
Few comments. 

We have an empty file that might be a left over.

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    const int resp_header = 2;
+    const int resp_payload = 4;
```
Comment (espressif-abhikroy) [inline components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp:71]: Will this work any better?
```suggestion
  enum class Access {
    request = 1,
    resp_header = 2,
    resp_payload = 4
  };
```

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_server.cpp (hunk 1 above)
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_server.cpp:34]: will add few doxygen comments, but this isn't a public API, so I'll keep it brief

File: components/esp_wifi_remote/eppp/eppp_init.c [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+#if CONFIG_ESP_WIFI_REMOTE_EPPP_TRANSPORT_UART
+    config.transport = EPPP_TRANSPORT_UART;
+    config.uart.tx_io = CONFIG_ESP_WIFI_REMOTE_EPPP_UART_TX_PIN;
+    config.uart.rx_io = CONFIG_ESP_WIFI_REMOTE_EPPP_UART_RX_PIN;
+#else
+#error ESP_WIFI_REMOTE supports only UART transport
+#endif
+    return eppp_open(role, &config, portMAX_DELAY);
+
+//    if (role == EPPP_SERVER) {
```
Comment (alisitsyn) [inline components/esp_wifi_remote/eppp/eppp_init.c]: Remove ?

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_server.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+#include "lwip/apps/snmp.h"
+
+extern "C" esp_netif_t *wifi_remote_eppp_init(eppp_type_t role);
+
+namespace eppp_rpc {
+
+namespace server {
+const char *TAG = "rpc_server";
+
+const unsigned char ca_crt[] = "-----BEGIN CERTIFICATE-----\n" CONFIG_ESP_WIFI_REMOTE_EPPP_CLIENT_CA "\n-----END CERTIFICATE-----";
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_server.cpp:26]: this version uses only compile time config, so the certs and keys are configured together with rpc options in kconfig only.

File: components/esp_wifi_remote/CMakeLists.txt [hunk 7]
```diff
@@ -1,15 +1,31 @@
...
+#endif()
+
 idf_component_register(INCLUDE_DIRS include
-                    SRCS  ${src_wifi_is_remote}
-                          esp_wifi_remote_net.c
-                          esp_wifi_remote_weak.c
+                    SRCS ${wifi_remote_host_srcs}
+                        ${src_wifi_remote_weak}
+                        eppp/wifi_remote_rpc_defs.c
+#                        eppp/wifi_remote_impl.cpp
```
Comment (alisitsyn) [inline components/esp_wifi_remote/CMakeLists.txt]: Is it kept for future update? Also this file contains the #if 0 #endif 
Suggest to remove if not needed now.

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_server.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_server.cpp:26]: follow-up task: Add an option to supply keys via a *global symbol*

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp (hunk 2 above)
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp:69]: yes, I'll use class Sync instead of struct and move some mebers to the private section

File: components/esp_wifi_remote/wifi_apps/roaming_app/src/Kconfig.roaming
Comment (david-cermak) [reply components/esp_wifi_remote/wifi_apps/roaming_app/src/Kconfig.roaming:1]: I'll add comments that this is needed for the Kconfig to work (it's referenced from esp_wifi Konfig)

File: components/esp_wifi_remote/eppp/eppp_init.c [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+#include "esp_log.h"
+#include "esp_wifi.h"
+#include "eppp_link.h"
+
+esp_netif_t *wifi_remote_eppp_init(eppp_type_t role)
+{
+    uint32_t our_ip = role == EPPP_SERVER ? EPPP_DEFAULT_SERVER_IP() : EPPP_DEFAULT_CLIENT_IP();
+    uint32_t their_ip = role == EPPP_SERVER ? EPPP_DEFAULT_CLIENT_IP() : EPPP_DEFAULT_SERVER_IP();
+    eppp_config_t config = EPPP_DEFAULT_CONFIG(our_ip, their_ip);
+#if CONFIG_ESP_WIFI_REMOTE_EPPP_TRANSPORT_UART
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/eppp_init.c:15]: Iwill remove the config and make the uart default and the only option

File: components/esp_wifi_remote/eppp/eppp_init.c [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    config.transport = EPPP_TRANSPORT_UART;
+    config.uart.tx_io = CONFIG_ESP_WIFI_REMOTE_EPPP_UART_TX_PIN;
+    config.uart.rx_io = CONFIG_ESP_WIFI_REMOTE_EPPP_UART_RX_PIN;
+#else
+#error ESP_WIFI_REMOTE supports only UART transport
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/eppp_init.c:20]: I'll add runtime configuration at later point, not this version. This would probably be a weak function, that users could override.

File: components/esp_wifi_remote/examples/mqtt/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,23 @@
+# This project serves as a demo to enable using esp-mqtt on ESP platform targets as well as on linux
+cmake_minimum_required(VERSION 3.16)
+
+if("${IDF_TARGET}" STREQUAL "linux")
+# For linux-target we have two options:
+# - With lwIP (must be defined on command line, e.g. idf.py -DWITH_LWIP=1)
+#     access networking from linux `tap` interface (TAP networking mode)
+# - Without lwIP (must be defined on command line, e.g. idf.py -DWITH_LWIP=0)
+#     no designated interface, accesses user network via linux/socket sys calls
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/examples/mqtt/CMakeLists.txt:9]: we should default to the option without lwip at this level of mqtt example (with lwip is too complex)

[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/examples/mqtt/CMakeLists.txt:9]: I'll cleanup the CMakeLists.txt

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+        ESP_RETURN_ON_ERROR(sync.notify(sync.request), TAG, "failed to notify req");
+        pending_resp = id;
+        ESP_RETURN_ON_ERROR(rpc.send(id), TAG, "Failed to send request");
+        return ESP_OK;
+    }
+
+    template<typename T>
+    T get_resp(api_id id)
+    {
+        sync.wait_for(sync.resp_header);
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp:98]: I'll leave it as is, since it's only used in two places next to ech other

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+        auto instance = static_cast<RpcInstance *>(ctx);
+        while (instance->perform() == ESP_OK) {}
+        vTaskDelete(nullptr);
+    }
+    esp_netif_t *netif{nullptr};
+};
+
+
+namespace client {
+RpcInstance instance;
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp:179]: i'll try to employ it and use it

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_impl.hpp [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#pragma once
+#include <cstring>
+#include <cerrno>
+
+namespace eppp_rpc {
+
+const int rpc_port = 3333;
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_impl.hpp:12]: will apply

[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_impl.hpp:12]: make sense. static is needed as it's being include to mulitple compilation units

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_impl.hpp (hunk 3 above)
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_impl.hpp:44]: i'll stick to the current member function name, as the operator* would be confusing and might indicate that we're dereferencing the instance pointer ( rpc->value() )

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp (hunk 4 above)
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp:71]: ok, will do.

File: components/esp_wifi_remote/eppp/eppp_init.c (hunk 5 above)
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/eppp_init.c]: will do

File: components/esp_wifi_remote/CMakeLists.txt (hunk 7 above)
Comment (david-cermak) [reply components/esp_wifi_remote/CMakeLists.txt]: already fixed in the current version

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_impl.hpp [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+        ESP_LOG_BUFFER_HEXDUMP("rpc", buf, size, ESP_LOG_VERBOSE);
+        int len = esp_tls_conn_write(tls_, buf, size);
+        if (len <= 0) {
+            ESP_LOGE("rpc", "Failed to write data to the connection");
+            return ESP_FAIL;
+        }
+        return ESP_OK;
+    }
+
+    esp_err_t send(api_id id) // specialization for (void)
```
Comment (david-cermak) [inline components/esp_wifi_remote/eppp/wifi_remote_rpc_impl.hpp:93]:
```suggestion
    esp_err_t send(api_id id) // overload of the templated method for (void)
```

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_server.cpp (hunk 1 above)
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_server.cpp:35]: Added brief comments to the main classes and structs

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_server.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_server.cpp:26]: Used the config option for now and added a `TODO` directly in the code to provide a mean of injecting user keys via an external file embedded to the binary.

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp (hunk 2 above)
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp:76]: Made them private, but needed to *befriend* this class to the RPC engine, as these are used also outside.

File: components/esp_wifi_remote/eppp/eppp_init.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/eppp_init.c:20]: Have marked the `wifi_remote_eppp_init()` weak, so users can override the initialization part.

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp]: Used `constinit` but had to make a few other things `constexpr`

File: components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp (hunk 4 above)
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/wifi_remote_rpc_client.cpp:78]: I left the const variables here, as I need them as integers for the event group.

File: components/esp_wifi_remote/eppp/eppp_init.c (hunk 5 above)
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/eppp_init.c]: already applied

File: components/esp_wifi_remote/eppp/eppp_init.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/eppp_init.c]: Removed mentioning all other transport. left only the UART option.

**Review by espressif-abhikroy — APPROVED** [review]

**Review by alisitsyn — APPROVED** [review]

**Review by gabsuren — APPROVED** [review]

File: components/esp_wifi_remote/Kconfig.soc_wifi_caps.in [hunk 15]
```diff
@@ -227,3 +227,11 @@ endif # ESP32C6
...
 
 endif # ESP32H2
+
+if SLAVE_IDF_TARGET_ESP32P4
+
+    config SLAVE_SOC_WIFI_LIGHT_SLEEP_CLK_WIDTH
+        int
+        default 12
+
+endif # ESP32P4
```
Comment (mantriyogesh) [inline components/esp_wifi_remote/Kconfig.soc_wifi_caps.in:237]: When can the ESP32-P4 act as slave? \
I mean, P4 doesn't have Wi-Fi..

File: components/esp_wifi_remote/eppp/eppp_init.c [hunk 16]
```diff
@@ -0,0 +1,20 @@
+/*
+ * SPDX-FileCopyrightText: 2024 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "esp_log.h"
+#include "esp_wifi.h"
+#include "eppp_link.h"
+
+__attribute__((weak)) esp_netif_t *wifi_remote_eppp_init(eppp_type_t role)
```
Comment (mantriyogesh) [inline components/esp_wifi_remote/eppp/eppp_init.c:10]: Can we move this init of eppp in the eppp code? \
This way the esp_wifi_remote would be cleaner and agnosttic?

File: components/esp_wifi_remote/esp_wifi_with_remote.c [hunk 17]
```diff
@@ -267,26 +267,6 @@ esp_err_t esp_wifi_set_csi(_Bool en)
...
-esp_err_t esp_wifi_set_ant(const wifi_ant_config_t *config)
-{
-    return esp_wifi_remote_set_ant(config);
-}
-
-esp_err_t esp_wifi_get_ant(wifi_ant_config_t *config)
-{
-    return esp_wifi_remote_get_ant(config);
-}
-
```
Comment (mantriyogesh) [inline components/esp_wifi_remote/esp_wifi_with_remote.c:289]: Why are these removed?

File: components/esp_wifi_remote/eppp/eppp_init.c (hunk 16 above)
Comment (david-cermak) [reply components/esp_wifi_remote/eppp/eppp_init.c:10]: The eppp initialization doesn't belong to to the `eppp_link` component, maybe we could create a separate component from the eppp related sources? But I'd probably do it in a future versions...

File: components/esp_wifi_remote/esp_wifi_with_remote.c (hunk 17 above)
Comment (david-cermak) [reply components/esp_wifi_remote/esp_wifi_with_remote.c:289]: These have been deprecated by in the recent `esp-wifi` library

File: components/esp_wifi_remote/eppp/eppp_init.c (hunk 16 above)
Comment (mantriyogesh) [reply components/esp_wifi_remote/eppp/eppp_init.c:10]: maybe, eppp_link/app or something within eppp_link?

Comment (david-cermak) [reply components/esp_wifi_remote/eppp/eppp_init.c:10]: I'd ideally add the initialization code to the user app, but that's not according to the base requirement that the wifi code should run seamlessly on targets with or without wifi.

Comment (mantriyogesh) [reply components/esp_wifi_remote/eppp/eppp_init.c:10]: Sure..
